nvme_gw_server.py

```diff
@@ -0,0 +1,295 @@
+        self.logger.info({
+            f"Adding listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
+        })
+        try:
+            return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_listener(
+                self.client,
+                request.nqn,
+                request.trtype,
+                request.traddr,
+                request.trsvcid,
+                request.adrfam,
+            )
+            self.logger.info(f"Status of add listener: {return_string}")
+        except Exception as ex:
+            self.logger.error(f"Add Listener returned with error: \n {ex}")
+            context.set_code(grpc.StatusCode.INTERNAL)
+            context.set_details(f"{ex}")
+            return pb2.req_status()
+
+        return pb2.req_status(status=return_string)
+
+
+def serve(gw_config_filename):
+
+    nvme_config = nvme_gw_config.NVMeGWConfig(gw_config_filename)
+
+    enable_auth = nvme_config.getboolean("config", "enable_auth")
+    gateway_addr = nvme_config.get("config", "gateway_addr")
+    gateway_port = nvme_config.get("config", "gateway_port")
+
+    server_key = nvme_config.get("mtls", "server_key")
+    server_cert = nvme_config.get("mtls", "server_cert")
+    client_cert = nvme_config.get("mtls", "client_cert")
+
+    grpc_max_workers = int(nvme_config.get("config", "grpc_server_max_workers"))
+
+    if enable_auth:
+
+        # read in key and certificate
+        with open(server_key, "rb") as f:
+            private_key = f.read()
+        with open(server_cert, "rb") as f:
+            server_crt = f.read()
+        with open(client_cert, "rb") as f:
+            client_crt = f.read()
+
+        # create server credentials & set client root certificate & set require_client_auth to True
+        server_credentials = grpc.ssl_server_credentials(
+            private_key_certificate_chain_pairs=[(private_key, server_crt)],
+            root_certificates=client_crt,
+            require_client_auth=True,
+        )
+
+        # create server
+        server = grpc.server(
+            futures.ThreadPoolExecutor(max_workers=grpc_max_workers))
+        gw_service = GWService(nvme_config)
+        gw_service.start_spdk()
+        pb2_grpc.add_NVMEGatewayServicer_to_server(gw_service, server)
+
+        # add secure port using crendentials
+        server.add_secure_port("{}:{}".format(gateway_addr, gateway_port),
+                               server_credentials)
+    else:
+
+        # Authentication is not enabled
+        server = grpc.server(
+            futures.ThreadPoolExecutor(max_workers=grpc_max_workers))
+        gw_service = GWService(nvme_config)
+        gw_service.start_spdk()
+        pb2_grpc.add_NVMEGatewayServicer_to_server(gw_service, server)
+        server.add_insecure_port("{}:{}".format(gateway_addr, gateway_port))
+
+    server.start()
+    server.wait_for_termination()
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(prog="nvme_gw",
```

Comment:
```suggestion
    parser = argparse.ArgumentParser(prog="python3 ./nvme_gw_server",
```
although I don't see why these scripts couldn't be made executable (shebang, etc).